### PR TITLE
Sprint 2 Release

### PR DESCRIPTION
**Pull request by AdrianAlucard**
Finished work for sprint 2...grove pi functionality implemented. 
Patterns Used:
CoR
inner classes 
Iterator
Abstract Factory
Template
Singleton
Adapter

### CODE REVIEW

**Comment by niehausks** [comment]
All the patterns! Looks good

File: PublicComponentCode/ComponentCode/src/java/main/ccsu/edu/grovepicomponents/TemperatureAndHumiditySensor.java [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+import edu.ccsu.utility.CommonConstants;
+
+/***
+ * Temperature and humidity sensor that reads data from
+ * the GrovePi temp/humidity sensor and returns that data
+ *
+ */
+public class TemperatureAndHumiditySensor extends Sensor {
+
+	private List<TempAndHumidityData> sensorData;
```
Comment (caw13) [inline PublicComponentCode/ComponentCode/src/java/main/ccsu/edu/grovepicomponents/TemperatureAndHumiditySensor.java]: And now the real power of generics and visibility comes into play.  The list of sensorData (used by all concrete classes currently: TemperatureAndHumiditySensor, LightSensor) can be refactored to the parent Sensor abstract class.  You obviously still want the sensor data typed such that it is useful within the individual subclass sensor so rather than the parent class dictating what the type is specify it as a generic that the subclass will provide the type.  Note also here that your sensor data classes should be nested static classes rather than just inner classes as they don't need to be tied to the outer class they were created by.  So essentially you want this:
  >public class TemperatureAndHumiditySensor extends Sensor<TemperatureAndHumiditySensor.TempAndHumidityData>{

But wait the compiler gives me a problem if I try and give it a private class as the generic!  Turns out this is the exact reason the protected key word is a bit odd in terms of its visibility so that it doesn't just give access to subclasses, but also everything in the package.  It was created for situations like this where you want to keep the inner class protected, but allow it to be specified as a possible type of the parent class when using generics.  This is actually a rather beautiful use of all of those pattern elements :)

So to summarize, for your data classes on both classes those should be static protected classes, you should refactor the sensor data list to the Sensor class making it use Generics, for the child classes now they will extend Sensor specifying which data class they are binding to.

File: PublicComponentCode/ComponentCode/src/java/main/ccsu/edu/grovepicomponents/TemperatureAndHumiditySensor.java [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+	public void setNextSensor(Sensor nextSensor) throws IncompatibleSensorError {
+		if(nextSensor instanceof TemperatureAndHumiditySensor) {
+			this.nextSensor = nextSensor;
+		} else {
+			throw new IncompatibleSensorError("Sensor not compatible with sensor. Sensor chain can only be use other Sensors");
+		}
+	}
+    
+	@Override
+	public Iterator getIterator() {
```
Comment (caw13) [inline PublicComponentCode/ComponentCode/src/java/main/ccsu/edu/grovepicomponents/TemperatureAndHumiditySensor.java]: Good use of Iterator protecting visibility, but have the Iterator returned type bound.  Your getIterator method doesn't need to be public.  If it is only used internally so you want to bind it to the protected data type then just change your visibility here to either private or protected and bind it to the internal class.

File: PublicComponentCode/ComponentCode/src/java/main/ccsu/edu/grovepicomponents/TemperatureAndHumiditySensor.java [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+					 "DegressCelcius: " + this.celcius + "\n" +
+					 "Humidity Percentage: " + this.humidityValue + "\n" +
+					 "Date: " + this.date + 
+					 "\n******************\n";
+		 }
+	}
+	/**
+	 * Inner Class for iterating over tempAndHumidity data
+	 * */
+	private class TempAndHumidityIterator implements Iterator{
```
Comment (caw13) [inline PublicComponentCode/ComponentCode/src/java/main/ccsu/edu/grovepicomponents/TemperatureAndHumiditySensor.java]: Good use of inner classes and iterator pattern
